Cancellation of reassignments allowed if the resulting ISR is at min-ISR

sutambe opened the pull request:
Current implementation of cancellation is more restrictive as in if any of the brokers in the original replica list are down, cancellations are disallowed. That's overly restrictive if one of the original brokers is down for maintenance while there're enough replicas alive in the original list that could handle read and write at min-ISR.
 
This implementation raises an error in cancellation only if the resulting ISR consisting of a subset of original alive brokers will push the partition under min-ISR.

### Testing
1. Unit test included.
2. Testing completed in prod-lor1 kafka.cert2 cluster. Full test details: https://jira01.corp.linkedin.com:8443/browse/LIKAFKA-50665

groelofs commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -392,10 +391,36 @@ class ReassignPartitionsIntegrationTest extends ZooKeeperTestHarness {
+  }
+
+  @Test
+  def testAboveMinISRCancellationSucceeds(): Unit = {
```
Technically this is AtMinISR, right? Maybe rename to testAtOrAboveMin... ?

sutambe replied:
done

groelofs commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -2302,21 +2302,35 @@ class KafkaController(val config: KafkaConfig,
+            Right(replicaAssignment.originReplicas)
+          } else {
+            val deadOriginalReplicas = replicaAssignment.originReplicas.toSet.diff(controllerContext.liveBrokerIds);
+            if (deadOriginalReplicas.size >= 2) {
+              // If there are two or more dead replicas, ISR would fall below minISR, which we don't allow
```
This is still the hard-coded logic Krit flagged, right?  This would allow cancellations with one dead broker even if it were configured at RF == minISR, and it would disallow cancellations with two dead brokers even if RF == minISR + 3 (for example). What we'd want here is a check for (deadOriginalReplicas.size > RF - minISR), right?

sutambe replied:
Based on the discussion with Krit yesterday, a config to control this behavior would be adequate. We don’t allow topic creation with `minISR == RF` or `minISR == RF-1` because we want to be able to handle one planned and one unplanned failure all the times. So `RF - minISR >= 2` is true in all prod/ei use-cases. This is applicable to `RF==3` and `RF==4`.

Under some test scenarios, where `RF >= 5`, the config I'm going to add will allow cancellation if enough original replicas are on board. 

Does this look right to you?

groelofs replied:
Depending on another config when we've already got two feels kludgy and error-prone to me, especially given that its description is fairly complex and subtle. Can we not simply derive this value? KafkaConfig already has min.insync.replicas and default.replication.factor; aren't those effectively what CC checks? I know individual topics can override this, and it would be better to check the RF of the topic in question explicitly, but I don't know offhand how accessible that metadata is at this point in the code.

sutambe replied:
We've two combinations . `minISR==2, RF==4` in KAC/KRC. then we've `minISR==1, RF==3` everywhere else. so defaults in linkedin-kafka level are not the same. In fact, in linkedin-kafka `Defaults.DefaultReplicationFactor` is `1`. 

Either way, the `RF-minISR` is 2 in Linkedin clusters. I think the suggestion made was to have maximum flexibility and potentially cater to other scenarios such as dev clusters and change in future etc. Therefore, I prefer a direct control independent of the other parameters, which as you said could change per topic.

groelofs replied:
~~Let me put it this way:  suppose you do this and set the global default to 2 as you've indicated, and then at some future point someone creates a cluster with RF = minISR + 1. They're not even aware of this new config, and as a result, you and they have just reintroduced the leaderless-partitions bug for that cluster. That's why it's a bad approach.~~

~~In general, _every_ manual approach is a bad approach:  if it _can_ be automated, it _should_ be automated. It's not always feasible to do so, but in this case the logic is clearcut, and the controller has total knowledge of all of the relevant pieces of info, so let's reduce the likelihood of errors rather than increase it.~~

OK, I agree. My (now edited) concerns above were based on a misreading of the new logic, which is safer than what I pictured. Resolving...

sutambe committed:
raise an error in cancellation only if the resulting ISR consisting of a subset of original alive brokers will push the partition under minISR
adding a configuration for liNumMinOriginalAliveReplicas with the same default value as that of minISR, which is 1
add config validation.

groelofs commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -2302,21 +2302,34 @@ class KafkaController(val config: KafkaConfig,
+            Right(replicaAssignment.originReplicas)
+          } else {
+            val aliveOriginalReplicas = replicaAssignment.originReplicas.toSet.intersect(controllerContext.liveBrokerIds);
+            if (aliveOriginalReplicas.size >= config.liNumMinOriginalAliveReplicas) {
+              // If there are enough original replicas alive (to maintain minISR), we allow the cancellation to go through
+              Right(aliveOriginalReplicas.toSeq)
+            } else {
+              // If there is only 1 replica from the original set, we don't allow the cancellation to go though.
+              Left(new ApiError(Errors.INVALID_REPLICA_ASSIGNMENT,
+                s"Replica assignment cancellation have 2 or more original brokers that are not alive. " +
+                  "Replica list: " + s"${replicaAssignment.originReplicas}, live broker list: ${controllerContext.liveBrokerIds}"))
```
nit: The two comments and the error message effectively hard-code the new config value. At least the error message should definitely change, or it will be super-confusing during debug sessions. Something like "Replica assignment cannot be cancelled because the number of original brokers that are still alive is less than the configured requirement (liNumMinAliveOriginalReplicas = ${config.liNumMinAliveOriginalReplicas})." with the replica list and live-broker lists attached, perhaps?

sutambe replied:
updated the comments.

groelofs commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -566,6 +567,7 @@ object KafkaConfig {
   val InterBrokerProtocolVersionProp = "inter.broker.protocol.version"
   val InterBrokerListenerNameProp = "inter.broker.listener.name"
   val ReplicaSelectorClassProp = "replica.selector.class"
+  val LiNumMinOriginalAliveReplicasProp = "li.num.min.original.alive.replicas"
```
nit: the "Num"/"num" part is unnecessary (e.g., "min.insync.replicas") and can be dropped, but if present, it should come after the "min" (i.e., "minimum number" not "number minimum").

sutambe replied:
fixed

groelofs commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -392,10 +391,36 @@ class ReassignPartitionsIntegrationTest extends ZooKeeperTestHarness {
+  }
+
+  @Test
+  def testBelowMinISRCancellationFails(): Unit = {
+    cluster = new ReassignPartitionsTestCluster(zkConnect)
+    val assignment = setupPartitionMovementBeforeCancellation(cluster)
+
+    // shutdown two of the original replicas. Cancellation would cause ISR to drop below minISR
+    cluster.servers.find(_.config.brokerId == 1).get.shutdown()
+    cluster.servers.find(_.config.brokerId == 2).get.shutdown()
+
+    // Cancel the reassignment should result in an exception because
+    // there are not enough alive replicas in the original list {0}
     assertThrows(classOf[TerseReassignmentFailureException], () => runCancelAssignment(cluster.adminClient, assignment, true))
```
I don't see your new config in here, and with its default value of 1, shouldn't this test case fail?

sutambe replied:
you are right. Updated the test.

sutambe committed:
updates

groelofs commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -2322,9 +2322,9 @@ class KafkaController(val config: KafkaConfig,
+              // If there are not enough alive replicas in the original set, we don't allow the cancellation to go though.
               Left(new ApiError(Errors.INVALID_REPLICA_ASSIGNMENT,
-                s"Replica assignment cancellation have 2 or more original brokers that are not alive. " +
+                s"Replica assignment cancellation precondition failed. Don't have enough alive replicas in the original replica set: " +
```
It would be good to include the config value, too, in case it's a different value than what the debugging dev thinks. There's no easy way to figure it out after the fact if logs have rotated, and not being aware of it can lead to wasted time.

sutambe replied:
updated the error message

sutambe committed:
updates

sutambe committed:
updates

groelofs reviewed (approved):
Thanks!
